src/ndarray/autograd.cc

```diff
@@ -0,0 +1,218 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file autograd.cc
+ * \brief Implementation of AutogradRuntime module.
+ */
+
+#include <mxnet/operator.h>
+#include <mxnet/executor.h>
+#include <nnvm/pass_functions.h>
+#include <unordered_set>
+#include <iostream>
+#include "../executor/graph_executor.h"
+#include "./autograd.h"
+
+namespace mxnet {
+namespace autograd {
+
+using nnvm::Symbol;
+using nnvm::Node;
+using nnvm::NodePtr;
+using nnvm::NodeEntry;
+using nnvm::NodeEntryMap;
+using exec::GraphExecutor;
+
+AutogradRuntime::AutogradRuntime() {}
+
+void AutogradRuntime::SetRecording(bool recording) {
+  is_recording_ = recording;
+}
+
+bool AutogradRuntime::IsRecording() const {
+  return is_recording_;
+}
+
+void AutogradRuntime::MarkVariables(std::vector<NDArray*> *p_variables) {
+  std::vector<NDArray*>& variables = *p_variables;
+  for (NDArray* var : variables) {
+    NodeEntry& e = var->entry_;
+    e.node = Node::Create();
+    e.node->attrs.name = "ag_variables_" + std::to_string(variable_count_++);
+  }
+}
+
+void AutogradRuntime::RecordImperativeFCompute(FCompute fn,
+                                               const nnvm::Op* op,
+                                               const nnvm::NodeAttrs& attrs,
+                                               std::vector<NDArray> *p_inputs,
+                                               std::vector<NDArray> *p_outputs) {
+  RecordOp(op, attrs, p_inputs, p_outputs);
+}
+
+void AutogradRuntime::RecordImperativeOperator(std::shared_ptr<Operator> opr,
+                                               const nnvm::Op* op,
+                                               const nnvm::NodeAttrs& attrs,
+                                               std::vector<NDArray> *p_inputs,
+                                               std::vector<NDArray> *p_outputs) {
+  NodePtr node = RecordOp(op, attrs, p_inputs, p_outputs);
+  saved_opr_.insert({node.get(), opr});
+}
+
+std::vector<NDArray> Execute(Symbol sym,
+                             const NodeEntryMap<NDArray>& feed_dict,
+                             const NodeOperatorMap& saved_opr);
+
+std::vector<NDArray> AutogradRuntime::ComputeGradient(const std::vector<NDArray>& outputs) {
+  Symbol ff_sym;
+  for (size_t i = 0; i < outputs.size(); ++i) {
+    ff_sym.outputs.push_back(outputs[i].entry_);
+  }
+  std::vector<NDArray> result = Execute(ff_sym, saved_ndarray_, saved_opr_);
+  ClearRecords();
```

Comment:
The clear records approach has some problems. for example you can only differentiate once and everything is destroyed. It maybe better to track this as reference counters in NDArray{ shared_ptr<AutogradEntry> };